pkg/api/dashboard.go

```diff
@@ -255,6 +258,225 @@ func GetDashboardFromJsonFile(c *middleware.Context) {
+		CreatedBy:        creator,
+	}
+
+	return Json(200, dashVersionMeta)
+}
+
+func createCompareDashboardVersionCommand(c *middleware.Context) (m.CompareDashboardVersionsCommand, error) {
+	cmd := m.CompareDashboardVersionsCommand{}
+
+	dashboardIdStr := c.Params(":dashboardId")
+	dashboardId, err := strconv.Atoi(dashboardIdStr)
+	if err != nil {
+		return cmd, err
+	}
+
+	versionStrings := strings.Split(c.Params(":versions"), "...")
+	if len(versionStrings) != 2 {
+		return cmd, fmt.Errorf("bad format: urls should be in the format /versions/0...1")
+	}
+
+	originalDash, err := strconv.Atoi(versionStrings[0])
+	if err != nil {
+		return cmd, fmt.Errorf("bad format: first argument is not of type int")
+	}
+
+	newDash, err := strconv.Atoi(versionStrings[1])
+	if err != nil {
+		return cmd, fmt.Errorf("bad format: second argument is not of type int")
+	}
+
+	cmd.DashboardId = int64(dashboardId)
+	cmd.Original = originalDash
+	cmd.New = newDash
+	return cmd, nil
+}
+
+// CompareDashboardVersions compares dashboards the way the GitHub API does.
+func CompareDashboardVersions(c *middleware.Context) Response {
+	cmd, err := createCompareDashboardVersionCommand(c)
+	if err != nil {
+		return ApiError(500, err.Error(), err)
+	}
+	cmd.DiffType = m.DiffDelta
+
+	if err := bus.Dispatch(&cmd); err != nil {
+		return ApiError(500, "Unable to compute diff", err)
+	}
+
+	// here the output is already JSON, so we need to unmarshal it into a
+	// map before marshaling the entire response
+	deltaMap := make(map[string]interface{})
+	err = json.Unmarshal(cmd.Delta, &deltaMap)
+	if err != nil {
+		return ApiError(500, err.Error(), err)
+	}
+
+	return Json(200, util.DynMap{
+		"meta": util.DynMap{
+			"original": cmd.Original,
+			"new":      cmd.New,
+		},
+		"delta": deltaMap,
+	})
+}
+
+// CompareDashboardVersionsJSON compares dashboards the way the GitHub API does,
+// returning a human-readable JSON diff.
+func CompareDashboardVersionsJSON(c *middleware.Context) Response {
+	cmd, err := createCompareDashboardVersionCommand(c)
+	if err != nil {
+		return ApiError(500, err.Error(), err)
+	}
+	cmd.DiffType = m.DiffJSON
+
+	if err := bus.Dispatch(&cmd); err != nil {
+		return ApiError(500, err.Error(), err)
+	}
+
+	return Respond(200, cmd.Delta).Header("Content-Type", "text/html")
+}
+
+// CompareDashboardVersionsBasic compares dashboards the way the GitHub API does,
+// returning a human-readable diff.
+func CompareDashboardVersionsBasic(c *middleware.Context) Response {
+	cmd, err := createCompareDashboardVersionCommand(c)
+	if err != nil {
+		return ApiError(500, err.Error(), err)
+	}
+	cmd.DiffType = m.DiffBasic
+
+	if err := bus.Dispatch(&cmd); err != nil {
+		return ApiError(500, err.Error(), err)
+	}
+
+	return Respond(200, cmd.Delta).Header("Content-Type", "text/html")
+}
+
+// RestoreDashboardVersion restores a dashboard to the given version.
+func RestoreDashboardVersion(c *middleware.Context, cmd m.RestoreDashboardVersionCommand) Response {
```

Comment:
This whole sql command seemed to duplicate a lof of existing code (Get dashboard, Get dashboard version, Save dashboard & insert history). 

I refactored this to use existing queries & commands (this also made restore work for alerting, so old versions of alert rules are now also restored, as the restore logic is using the exact same code path as any dashboard save instead of duplicating it like was done below). 
https://github.com/grafana/grafana/commit/c87418d060e932583b053bcb8ac9e53161b1141c